File: Headless/internal/server/panegroup.go
package server

import (
	"errors"
	"fmt"
	"sort"
	"strings"
	"time"

	"github.com/abcdlsj/warren/Headless/internal/api"
	"github.com/abcdlsj/warren/Headless/internal/pane"
	"github.com/abcdlsj/warren/Headless/internal/store"
)

// Pane Groups are durable Host state: one whole-screen arrangement of running
// Sessions per group, any number of groups per Workspace or Terminal Group.
//
// The rules live here rather than in a client because the Host is the only
// authority both the arrangement and the Sessions can be validated against. A
// client stores only which group it is looking at and which pane has focus.
//
// Error messages carry a stable prefix so a script can tell the cases apart
// without parsing prose:
//
//	pane group not found: <id>
//	pane group revision conflict: ...
//	pane group invalid tree: ...
//	pane group session not found: <id>
//	pane group session in use: <id>
//	pane group limit reached: ...
const (
	errPaneGroupNotFound     = "pane group not found"
	errPaneGroupRevConflict  = "pane group revision conflict"
	errPaneGroupInvalidTree  = "pane group invalid tree"
	errPaneGroupSessionGone  = "pane group session not found"
	errPaneGroupSessionInUse = "pane group session in use"
	errPaneGroupLimit        = "pane group limit reached"
)

// PaneGroups returns the groups of one owner in draw order. The CLI reads the
// roster instead; this is for the Host's own projections and tests.
func (s *Service) PaneGroups(ownerID string) []api.PaneGroup {
	state := s.Store.Snapshot()
	result := make([]api.PaneGroup, 0, len(state.PaneGroups))
	for _, group := range state.PaneGroups {
		if group.OwnerID() == ownerID {
			result = append(result, group)
		}
	}
	sortPaneGroups(result)
	return result
}

// PaneGroup returns one group by ID.
func (s *Service) PaneGroup(id string) (api.PaneGroup, bool) {
	for _, group := range s.Store.Snapshot().PaneGroups {
		if group.ID == id {
			return group, true
		}
	}
	return api.PaneGroup{}, false
}

// CreatePaneGroup opens a single-pane arrangement around an existing Session.
// A group never exists without a pane: an empty arrangement would have nothing
// to render and nothing to own.
func (s *Service) CreatePaneGroup(ownerScope, ownerID, sessionID, name, before string) (api.PaneGroup, error) {
	name = strings.TrimSpace(name)
	if len(name) > pane.MaxNameLength {
		return api.PaneGroup{}, fmt.Errorf("%s: name is longer than %d characters", errPaneGroupInvalidTree, pane.MaxNameLength)
	}
	var created api.PaneGroup
	err := s.Store.Update(func(value *api.State) error {
		normalizedScope, err := paneGroupOwner(value, ownerScope, ownerID)
		if err != nil {
			return err
		}
		if err := validatePaneGroupSession(value, sessionID, normalizedScope, ownerID, ""); err != nil {
			return err
		}
		if count := len(paneGroupsOf(value, ownerID)); count >= pane.MaxGroupsPerOwner {
			return fmt.Errorf("%s: %s already holds %d groups", errPaneGroupLimit, ownerID, count)
		}
		now := time.Now().UTC()
		group := api.PaneGroup{
			ID:        store.NewID(),
			Scope:     normalizedScope,
			Name:      name,
			Tree:      api.PaneNode{PaneID: store.NewID(), SessionID: sessionID},
			Revision:  1,
			CreatedAt: now,
			UpdatedAt: now,
		}
		if normalizedScope == api.SessionScopeTerminalGroup {
			group.TerminalGroupID = ownerID
		} else {
			group.WorkspaceID = ownerID
		}
		value.PaneGroups = append(value.PaneGroups, group)
		if err := reorderPaneGroups(value, group.ID, before); err != nil {
			return err
		}
		created = value.PaneGroups[indexOfPaneGroup(value, group.ID)]
		return nil
	})
	if err != nil {
		return api.PaneGroup{}, err
	}
	return created, nil
}

// RenamePaneGroup sets the optional user label. A rename is not a tree change,
// so it leaves Revision alone: a divider drag in flight stays valid.
func (s *Service) RenamePaneGroup(id, name string) (api.PaneGroup, error) {
	name = strings.TrimSpace(name)
	if len(name) > pane.MaxNameLength {
		return api.PaneGroup{}, fmt.Errorf("%s: name is longer than %d characters", errPaneGroupInvalidTree, pane.MaxNameLength)
	}
	var renamed api.PaneGroup
	err := s.Store.Update(func(value *api.State) error {
		index := indexOfPaneGroup(value, id)
		if index < 0 {
			return fmt.Errorf("%s: %s", errPaneGroupNotFound, id)
		}
		value.PaneGroups[index].Name = name
		value.PaneGroups[index].UpdatedAt = time.Now().UTC()
		renamed = value.PaneGroups[index]
		return nil
	})
	if err != nil {
		return api.PaneGroup{}, err
	}
	return renamed, nil
}

// MovePaneGroup reorders one group among its owner's groups. Other owners and
// other groups keep their identity, revision, and tree.
func (s *Service) MovePaneGroup(id, before string) (api.PaneGroup, error) {
	var moved api.PaneGroup
	err := s.Store.Update(func(value *api.State) error {
		if indexOfPaneGroup(value, id) < 0 {
			return fmt.Errorf("%s: %s", errPaneGroupNotFound, id)
		}
		if err := reorderPaneGroups(value, id, before); err != nil {
			return err
		}
		moved = value.PaneGroups[indexOfPaneGroup(value, id)]
		return nil
	})
	if err != nil {
		return api.PaneGroup{}, err
	}
	return moved, nil
}

// RemovePaneGroup deletes an arrangement. The Sessions it showed keep running
// and stay reachable as ordinary Tabs: a layout edit never ends a process.
func (s *Service) RemovePaneGroup(id string) error {
	if err := s.Store.Update(func(value *api.State) error {
		index := indexOfPaneGroup(value, id)
		if index < 0 {
			return fmt.Errorf("%s: %s", errPaneGroupNotFound, id)
		}
		value.PaneGroups = append(value.PaneGroups[:index:index], value.PaneGroups[index+1:]...)
		return nil
	}); err != nil {
		return err
	}
	return nil
}

// UpdatePaneGroup replaces one group's tree. This is the only structural
// mutator: splitting, closing a pane, placing a Session into a pane, and
// dragging a divider are all one tree replacement, guarded by the revision the
// caller observed.
//
// An incoming leaf without a Pane ID is assigned one here, because pane
// identity belongs to the Host: two clients that split at the same moment must
// not be able to invent the same identity.
func (s *Service) UpdatePaneGroup(id string, tree api.PaneNode, expectedRevision uint64) (api.PaneGroup, error) {
	var updated api.PaneGroup
	err := s.Store.Update(func(value *api.State) error {
		index := indexOfPaneGroup(value, id)
		if index < 0 {
			return fmt.Errorf("%s: %s", errPaneGroupNotFound, id)
		}
		group := value.PaneGroups[index]
		if expectedRevision != group.Revision {
			return fmt.Errorf("%s: group %s is at revision %d, caller expected %d; refresh the roster and retry",
				errPaneGroupRevConflict, id, group.Revision, expectedRevision)
		}
		candidate := clonePaneNode(&tree)
		pane.Normalize(candidate)
		pane.AssignPaneIDs(candidate, store.NewID)
		if err := pane.ValidateStructure(candidate); err != nil {
			return fmt.Errorf("%s: %w", errPaneGroupInvalidTree, err)
		}
		for _, leaf := range pane.Leaves(candidate) {
			if err := validatePaneGroupSession(value, leaf.SessionID, group.OwnerScope(), group.OwnerID(), id); err != nil {
				return err
			}
		}
		group.Tree = *candidate
		group.Revision++
		group.UpdatedAt = time.Now().UTC()
		value.PaneGroups[index] = group
		updated = group
		return nil
	})
	if err != nil {
		return api.PaneGroup{}, err
	}
	return updated, nil
}

// reconcilePaneGroups prunes every arrangement against the state's own Sessions
// and owners. It is the only reconciler, so a leaf can never disagree with the
// Session it shows on the wire, in state.json, or in a renderer.
//
// It reports whether anything changed, because the caller must not rewrite the
// state file once per tick for a steady Host.
func reconcilePaneGroups(state *api.State) bool {
	if len(state.PaneGroups) == 0 {
		return false
	}
	live := make(map[string]api.Session, len(state.Sessions))
	for _, session := range state.Sessions {
		if session.Lifecycle == "running" {
			live[session.ID] = session
		}
	}
	workspaces := make(map[string]struct{}, len(state.Workspaces))
	for _, workspace := range state.Workspaces {
		workspaces[workspace.ID] = struct{}{}
	}
	groups := make(map[string]struct{}, len(state.TerminalGroups))
	for _, group := range state.TerminalGroups {
		groups[group.ID] = struct{}{}
	}
	ordered := make([]api.PaneGroup, len(state.PaneGroups))
	copy(ordered, state.PaneGroups)
	sortPaneGroups(ordered)

	claimed := make(map[string]string, len(live))
	kept := make([]api.PaneGroup, 0, len(ordered))
	changed := false
	for _, group := range ordered {
		ownerID := group.OwnerID()
		ownerGone := ownerID == ""
		if !ownerGone {
			if group.OwnerScope() == api.SessionScopeTerminalGroup {
				_, known := groups[ownerID]
				ownerGone = !known
			} else {
				_, known := workspaces[ownerID]
				ownerGone = !known
			}
		}
		if ownerGone {
			changed = true
			continue
		}
		tree := *clonePaneNode(&group.Tree)
		// A Session claimed by an earlier group is dropped here rather than
		// rendered twice: one arrangement owns a Session at a time.
		reconciled := pane.Reconcile(&tree, func(sessionID string) bool {
			session, ok := live[sessionID]
			if !ok {
				return false
			}
			if !sessionBelongsToOwner(session, group.OwnerScope(), ownerID) {
				return false
			}
			if _, taken := claimed[sessionID]; taken {
				return false
			}
			return true
		})
		if reconciled == nil {
			changed = true
			continue
		}
		for _, leaf := range pane.Leaves(reconciled) {
			claimed[leaf.SessionID] = group.ID
		}
		if !samePaneNode(reconciled, &group.Tree) {
			group.Tree = *reconciled
			group.Revision++
			group.UpdatedAt = time.Now().UTC()
			changed = true
		}
		kept = append(kept, group)
	}
	if !changed && len(kept) == len(state.PaneGroups) {
		return false
	}
	state.PaneGroups = kept
	return true
}

// paneGroupOwner validates the owner and returns its normalized scope.
func paneGroupOwner(state *api.State, ownerScope, ownerID string) (string, error) {
	ownerID = strings.TrimSpace(ownerID)
	if ownerID == "" {
		return "", errors.New("pane group requires a workspace or terminal group")
	}
	switch ownerScope {
	case api.SessionScopeWorkspace:
		if !workspaceExists(state, ownerID) {
			return "", fmt.Errorf("workspace not found: %s", ownerID)
		}
		return api.SessionScopeWorkspace, nil
	case api.SessionScopeTerminalGroup:
		if !terminalGroupExists(state, ownerID) {
			return "", fmt.Errorf("terminal group not found: %s", ownerID)
		}
		return api.SessionScopeTerminalGroup, nil
	default:
		return "", fmt.Errorf("unknown pane group owner scope %q", ownerScope)
	}
}

// validatePaneGroupSession checks that one Session may be shown by one group of
// this owner. ignoreGroup exempts the group being updated: its own panes are
// not "in use" by itself.
func validatePaneGroupSession(state *api.State, sessionID, ownerScope, ownerID, ignoreGroup string) error {
	session, ok := sessionOfState(state, sessionID)
	if !ok || session.Lifecycle != "running" {
		return fmt.Errorf("%s: %s", errPaneGroupSessionGone, sessionID)
	}
	if !sessionBelongsToOwner(session, ownerScope, ownerID) {
		return fmt.Errorf("pane group session %s does not belong to %s", sessionID, ownerID)
	}
	for _, group := range state.PaneGroups {
		if group.ID == ignoreGroup {
			continue
		}
		for _, leaf := range pane.Leaves(&group.Tree) {
			if leaf.SessionID == sessionID {
				return fmt.Errorf("%s: %s is already shown by pane group %s", errPaneGroupSessionInUse, sessionID, group.ID)
			}
		}
	}
	return nil
}

func sessionBelongsToOwner(session api.Session, ownerScope, ownerID string) bool {
	if ownerScope == api.SessionScopeTerminalGroup {
		return session.TerminalGroupID == ownerID
	}
	return session.WorkspaceID == ownerID
}

func sessionOfState(state *api.State, id string) (api.Session, bool) {
	for _, session := range state.Sessions {
		if session.ID == id {
			return session, true
		}
	}
	return api.Session{}, false
}

func paneGroupsOf(state *api.State, ownerID string) []api.PaneGroup {
	result := make([]api.PaneGroup, 0, len(state.PaneGroups))
	for _, group := range state.PaneGroups {
		if group.OwnerID() == ownerID {
			result = append(result, group)
		}
	}
	return result
}

func indexOfPaneGroup(state *api.State, id string) int {
	for index := range state.PaneGroups {
		if state.PaneGroups[index].ID == id {
			return index
		}
	}
	return -1
}

// reorderPaneGroups moves one group to sit before another of its owner, or to
// the end when before is empty, then renumbers the owner's Order values. Groups
// of other owners are never touched.
func reorderPaneGroups(state *api.State, id, before string) error {
	index := indexOfPaneGroup(state, id)
	if index < 0 {
		return fmt.Errorf("%s: %s", errPaneGroupNotFound, id)
	}
	group := state.PaneGroups[index]
	owner := group.OwnerID()
	peers := paneGroupsOf(state, owner)
	sortPaneGroups(peers)
	target := len(peers)
	if before != "" {
		found := false
		for position := range peers {
			if peers[position].ID == before {
				target = position
				found = true
				break
			}
		}
		if !found {
			return fmt.Errorf("pane group move target not found: %s", before)
		}
	}
	reordered := make([]api.PaneGroup, 0, len(peers))
	for _, peer := range peers {
		if peer.ID != id {
			reordered = append(reordered, peer)
		}
	}
	if target > len(reordered) {
		target = len(reordered)
	}
	reordered = append(reordered, api.PaneGroup{})
	copy(reordered[target+1:], reordered[target:])
	reordered[target] = group
	for position := range reordered {
		reordered[position].Order = position
	}
	rebuildPaneGroups(state, owner, reordered)
	return nil
}

// rebuildPaneGroups replaces one owner's groups in place while leaving every
// other owner's entries where they were.
func rebuildPaneGroups(state *api.State, ownerID string, updated []api.PaneGroup) {
	next := make([]api.PaneGroup, 0, len(state.PaneGroups))
	inserted := false
	for _, group := range state.PaneGroups {
		if group.OwnerID() != ownerID {
			next = append(next, group)
			continue
		}
		if !inserted {
			next = append(next, updated...)
			inserted = true
		}
	}
	state.PaneGroups = next
}

func sortPaneGroups(groups []api.PaneGroup) {
	sort.Slice(groups, func(i, j int) bool {
		if groups[i].Order != groups[j].Order {
			return groups[i].Order < groups[j].Order
		}
		if !groups[i].CreatedAt.Equal(groups[j].CreatedAt) {
			return groups[i].CreatedAt.Before(groups[j].CreatedAt)
		}
		return groups[i].ID < groups[j].ID
	})
}

// clonePaneNode deep-copies a tree so a caller's value and the stored state can
// never alias: Reconcile and Remove mutate nodes in place.
func clonePaneNode(node *api.PaneNode) *api.PaneNode {
	if node == nil {
		return nil
	}
	cloned := &api.PaneNode{
		PaneID:    node.PaneID,
		SessionID: node.SessionID,
		Axis:      node.Axis,
		Ratio:     node.Ratio,
	}
	cloned.First = clonePaneNode(node.First)
	cloned.Second = clonePaneNode(node.Second)
	return cloned
}

// samePaneNode compares two trees structurally, including pane identity, so a
// reconciliation can tell "nothing to do" from "the arrangement changed".
func samePaneNode(a, b *api.PaneNode) bool {
	if a == nil || b == nil {
		return a == nil && b == nil
	}
	if a.Axis != b.Axis || a.PaneID != b.PaneID || a.SessionID != b.SessionID {
		return false
	}
	if a.Axis == "" {
		return true
	}
	if a.Ratio != b.Ratio {
		return false
	}
	return samePaneNode(a.First, b.First) && samePaneNode(a.Second, b.Second)
}

File: Packages/Domain/Sources/WarrenDomain/PaneGroup.swift
import Foundation

/// Which way a split divides its container.
public enum SplitAxis: String, Codable, Hashable, Sendable {
    /// Left and right columns.
    case horizontal
    /// Top and bottom rows.
    case vertical
}

/// One node of a Pane Group's tree.
///
/// A leaf carries the identity of its pane and of the Session it shows; a split
/// carries geometry only. Pane identity belongs to the Host, so a leaf a client
/// is creating arrives without one and is filled in by the Host.
public indirect enum PaneNode: Hashable, Sendable {
    case leaf(paneID: PaneID?, sessionID: TerminalSessionID)
    case split(axis: SplitAxis, ratio: Double, first: PaneNode, second: PaneNode)

    public static func leaf(sessionID: TerminalSessionID) -> PaneNode {
        .leaf(paneID: nil, sessionID: sessionID)
    }

    public var paneID: PaneID? {
        switch self {
        case .leaf(let paneID, _): return paneID
        case .split: return nil
        }
    }

    public var sessionID: TerminalSessionID? {
        switch self {
        case .leaf(_, let sessionID): return sessionID
        case .split: return nil
        }
    }

    public var isLeaf: Bool {
        if case .leaf = self { return true }
        return false
    }

    /// Leaves in preorder: the order a client renders and numbers panes in.
    public var leaves: [PaneNode] {
        switch self {
        case .leaf: return [self]
        case .split(_, _, let first, let second): return first.leaves + second.leaves
        }
    }

    public var paneCount: Int { leaves.count }

    public var sessionIDs: [TerminalSessionID] { leaves.compactMap(\.sessionID) }

    public func paneIndex(forSession sessionID: TerminalSessionID) -> Int? {
        leaves.firstIndex { $0.sessionID == sessionID }.map { $0 + 1 }
    }

    /// Rebuilds the tree with one leaf's Session replaced, keeping pane identity
    /// and geometry. Returns nil when the Session is not in the tree.
    public func replacingSession(
        _ sessionID: TerminalSessionID,
        with replacement: TerminalSessionID
    ) -> PaneNode? {
        switch self {
        case .leaf(let paneID, let existing):
            guard existing == sessionID else { return nil }
            return .leaf(paneID: paneID, sessionID: replacement)
        case .split(let axis, let ratio, let first, let second):
            if let updated = first.replacingSession(sessionID, with: replacement) {
                return .split(axis: axis, ratio: ratio, first: updated, second: second)
            }
            if let updated = second.replacingSession(sessionID, with: replacement) {
                return .split(axis: axis, ratio: ratio, first: first, second: updated)
            }
            return nil
        }
    }

    /// Rebuilds the tree with one leaf's ratio replaced, addressed by the child
    /// path a divider drag owns: false descends to first, true to second.
    public func replacingRatio(_ ratio: Double, at path: [Bool]) -> PaneNode? {
        guard case .split(let axis, let existing, let first, let second) = self else { return nil }
        guard let firstStep = path.first else {
            return .split(axis: axis, ratio: ratio, first: first, second: second)
        }
        if firstStep {
            guard let updated = second.replacingRatio(ratio, at: Array(path.dropFirst())) else { return nil }
            return .split(axis: axis, ratio: existing, first: first, second: updated)
        }
        guard let updated = first.replacingRatio(ratio, at: Array(path.dropFirst())) else { return nil }
        return .split(axis: axis, ratio: existing, first: updated, second: second)
    }
}

extension PaneNode: Codable {
    private enum CodingKeys: String, CodingKey {
        case paneId, sessionId, axis, ratio, first, second
    }

    public init(from decoder: Decoder) throws {
        let values = try decoder.container(keyedBy: CodingKeys.self)
        if let axis = try values.decodeIfPresent(SplitAxis.self, forKey: .axis) {
            self = .split(
                axis: axis,
                ratio: try values.decodeIfPresent(Double.self, forKey: .ratio) ?? 0.5,
                first: try values.decode(PaneNode.self, forKey: .first),
                second: try values.decode(PaneNode.self, forKey: .second)
            )
            return
        }
        let rawSession = try values.decodeIfPresent(String.self, forKey: .sessionId) ?? ""
        guard let sessionID = TerminalSessionID(uuidString: rawSession) else {
            throw DecodingError.dataCorruptedError(
                forKey: .sessionId,
                in: values,
                debugDescription: "A pane leaf needs a Session ID."
            )
        }
        let rawPane = try values.decodeIfPresent(String.self, forKey: .paneId)
        self = .leaf(paneID: rawPane.flatMap(PaneID.init(uuidString:)), sessionID: sessionID)
    }

    public func encode(to encoder: Encoder) throws {
        var values = encoder.container(keyedBy: CodingKeys.self)
        switch self {
        case .leaf(let paneID, let sessionID):
            try values.encodeIfPresent(paneID?.description, forKey: .paneId)
            try values.encode(sessionID.description, forKey: .sessionId)
        case .split(let axis, let ratio, let firstChild, let secondChild):
            try values.encode(axis, forKey: .axis)
            try values.encode(ratio, forKey: .ratio)
            try values.encode(firstChild, forKey: .first)
            try values.encode(secondChild, forKey: .second)
        }
    }
}

/// One whole-screen arrangement of running Sessions, owned by the Host.
///
/// A Pane Group is durable Host state: the Host stores the tree, so any client
/// reads the same shape without a reporting peer, and several arrangements can
/// coexist in one Workspace or Terminal Group.
public struct PaneGroup: Identifiable, Codable, Hashable, Sendable {
    public let id: PaneGroupID
    public let hostID: HostID
    /// Workspace and Terminal Group ownership are mutually exclusive, exactly as
    /// they are for a Session.
    public let workspaceID: WorkspaceID?
    public let terminalGroupID: TerminalGroupID?
    public var name: String?
    public var order: Int
    public var tree: PaneNode
    /// The compare-and-swap token for a tree update. A rename or a reorder is
    /// not a tree change and leaves it alone.
    public var revision: UInt64
    public let createdAt: Date
    public var updatedAt: Date

    public init(
        id: PaneGroupID = PaneGroupID(),
        hostID: HostID,
        workspaceID: WorkspaceID? = nil,
        terminalGroupID: TerminalGroupID? = nil,
        name: String? = nil,
        order: Int = 0,
        tree: PaneNode,
        revision: UInt64 = 0,
        createdAt: Date = Date(),
        updatedAt: Date = Date()
    ) {
        self.id = id
        self.hostID = hostID
        self.workspaceID = workspaceID
        self.terminalGroupID = terminalGroupID
        self.name = name
        self.order = order
        self.tree = tree
        self.revision = revision
        self.createdAt = createdAt
        self.updatedAt = updatedAt
    }

    public var paneCount: Int { tree.paneCount }

    /// The Workspace or Terminal Group that owns the arrangement.
    public var ownerID: String { workspaceID?.description ?? terminalGroupID?.description ?? "" }

    /// A group is drawn as a group only once it holds more than one pane: a
    /// single pane has nothing to bind it to.
    public var isSplit: Bool { tree.paneCount > 1 }

    public func paneIndex(forSession sessionID: TerminalSessionID) -> Int? {
        tree.paneIndex(forSession: sessionID)
    }
}
